cw721-timelock/src/contract.rs

```diff
@@ -0,0 +1,272 @@
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    let ctx = ExecuteContext::new(deps, info, env);
+
+    match msg {
+        ExecuteMsg::AMPReceive(pkt) => {
+            ADOContract::default().execute_amp_receive(ctx, pkt, handle_execute)
+        }
+        _ => handle_execute(ctx, msg),
+    }
+}
+
+pub fn handle_execute(ctx: ExecuteContext, msg: ExecuteMsg) -> Result<Response, ContractError> {
+
+    match msg {
+        ExecuteMsg::ReceiveNft(msg) => handle_receive_cw721(ctx, msg),
+        ExecuteMsg::ClaimNft {
+            cw721_contract,
+            token_id,
+        } => execute_claim_cw721(ctx, cw721_contract, token_id),
+        _ => ADOContract::default().execute(ctx, msg),
+    }
+}
+
+fn handle_receive_cw721(ctx: ExecuteContext, msg: Cw721ReceiveMsg) -> Result<Response, ContractError> {
+
+    ADOContract::default().is_permissioned(
+        ctx.deps.storage,
+        ctx.env.clone(),
+        SEND_NFT_ACTION,
+        ctx.info.sender.clone(),
+    )?;
+
+    match from_json(&msg.msg)? {
+        Cw721HookMsg::TimelockNft {
+            lock_duration,
+            recipient,
+        } => execute_timelock_cw721(
+            ctx,
+            msg.sender,
+            msg.token_id,
+            lock_duration,
+            recipient,
+        ),
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+fn execute_timelock_cw721(
+    ctx: ExecuteContext,
+    _sender: String,
+    token_id: String,
+    lock_duration: MillisecondsDuration,
+    recipient: Recipient,
+) -> Result<Response<Empty>, ContractError> {
+    let ExecuteContext {
+        deps,
+        info,
+        env,
+        ..
+    } = ctx;
+
+    ensure!(
+        lock_duration.seconds() >= ONE_DAY,
+        ContractError::LockTimeTooShort {}
+    );
+    ensure!(
+        lock_duration.seconds() <= ONE_YEAR,
+        ContractError::LockTimeTooLong {}
+    );
+
+    let nft_address = info.sender.to_string();
+    let lock_id = format!("{}:{}", nft_address, token_id);
+
+    let recipient_addr = AndrAddr::from_string(recipient.get_addr()).get_raw_address(&deps.as_ref())?;
+    let timelock_info = TimelockInfo {
+        unlock_time: Milliseconds::from_seconds(env.block.time.seconds() + lock_duration.seconds()),
+        recipient: recipient_addr,
+    };
+
+    TIMELOCKS.save(deps.storage, lock_id.as_str(), &timelock_info)?;
+
+    Ok(Response::new()
+        .add_attributes(vec![
+            attr("method", "timelock_cw721"),
+            attr("lock_id", lock_id),
+        ])
+    )
+}
+
+fn execute_claim_cw721(
+    ctx: ExecuteContext,
+    cw721_contract: AndrAddr,
+    token_id: String,
+) -> Result<Response<Empty>, ContractError> {
+    let ExecuteContext {
+        deps,
+        env,
+        ..
+    } = ctx;
+
+    let lock_id = format!("{}:{}", cw721_contract.get_raw_address(&deps.as_ref())?, token_id);
+    let timelock_info: Option<TimelockInfo> = Some(TIMELOCKS.load(deps.storage, lock_id.as_str())?);
```

Comment:
Your error is gonna be gobbled here with the type conversion. Either use `map_err` to map it to `ContractError::NFTNotFound` or use `.may_load` and let your conditional on line 205 catch it. 

I'd suggest writing a unit test for this case (claiming a token that hasn't been locked).